### PR TITLE
Add deposit index to beacon state

### LINKED ISSUE #2199: Regular Sync Inserts PostChainStart Deposits Again Into the State
This is part of #1586. We have a major bug in which after restarting a node, regular sync processes previous post-chainstart deposits again into the state, leading validators to double their money. We should not allow this to be the case.

### PR DESCRIPTION
**Pull request by 0xKiwi**
Resolves #2199 

Currently, we have an issue that we don't properly make sure deposits aren't duplicated. The spec has recently implemented a deposit_index in the beacon state that maintains what deposit index has been processed for the sake of ensuring duplicate deposits aren't sent in. 

Changes made:

Beacon State now has DepositIndex which is:

* Initialized as 0.
* Incremented with each deposit.
* Used as the reference for pruning pending deposits.

MerkleBranchHash32S -> MerkleProofHash32S

### CODE REVIEW

File: beacon-chain/sync/initial-sync/sync_state.go [hunk 1]
```diff
@@ -47,17 +46,7 @@ func (s *InitialSync) processState(msg p2p.Message) {
 		return
 	}
 
-	exists, blkNum, err := s.powchain.BlockExists(ctx, bytesutil.ToBytes32(finalizedState.LatestEth1Data.BlockHash32))
```
Comment (prestonvanloon) [inline beacon-chain/sync/initial-sync/sync_state.go]: Why remove this? dont we still want to verify that the block exists?

File: beacon-chain/core/blocks/block_operations.go [hunk 2]
```diff
@@ -621,18 +621,29 @@ func ProcessValidatorDeposits(
...
+		beaconState.DepositIndex++
 	}
 	return beaconState, nil
 }
 
 func verifyDeposit(beaconState *pb.BeaconState, deposit *pb.Deposit) error {
+	// Deposits must be processed in order
+	if deposit.MerkleTreeIndex != beaconState.DepositIndex {
+		return fmt.Errorf(
+			"expected deposit merkle branch index to match beacon state deposit index, wanted: %d, received: %d",
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/block_operations.go]: Merkle branch index or merkle tree index?

File: beacon-chain/rpc/beacon_server.go [hunk 3]
```diff
@@ -244,6 +243,16 @@ func (bs *BeaconServer) PendingDeposits(ctx context.Context, _ *ptypes.Empty) (*
 		return nil, fmt.Errorf("could not generate historical deposit trie from deposits: %v", err)
 	}
 
+	allPendingDeps := bs.beaconDB.PendingDeposits(ctx, bNum)
```
Comment (prestonvanloon) [inline beacon-chain/rpc/beacon_server.go]: Is this sorted? If it's not sorted then your logic below will not work. 

I would not count on this being sorted.

Comment (0xKiwi) [reply beacon-chain/rpc/beacon_server.go]: It's sorted in the PendingDeposits function by merkle index.

File: beacon-chain/sync/initial-sync/sync_state.go (hunk 1 above)
Comment (0xKiwi) [reply beacon-chain/sync/initial-sync/sync_state.go]: Good catch, adding it back.

File: beacon-chain/core/blocks/block_operations.go (hunk 2 above)
Comment (0xKiwi) [reply beacon-chain/core/blocks/block_operations.go]: Good catch, thank you. Changed to merkle tree index.

File: beacon-chain/db/pending_deposits.go [hunk 4]
```diff
@@ -113,7 +113,7 @@ func (db *BeaconDB) PrunePendingDeposits(ctx context.Context, b *big.Int) {
 	// added deposit and works backwards.
 	idx := -1
 	for i := depositListSize - 1; i >= 0; i-- {
-		if db.pendingDeposits[i].block.Cmp(b) == -1 {
+		if db.pendingDeposits[i].deposit.MerkleTreeIndex < merkleTreeIndex {
```
Comment (nisdas) [inline beacon-chain/db/pending_deposits.go]: this relies on the pending deposits being sorted by their indexes, right now they are sorted by the block in which the deposit is logged in

File: beacon-chain/core/state/state.go [hunk 5]
```diff
@@ -146,6 +147,8 @@ func GenesisBeaconState(
 		if err != nil {
 			return nil, fmt.Errorf("could not process validator deposit: %v", err)
 		}
+
+		state.DepositIndex++
```
Comment (nisdas) [inline beacon-chain/core/state/state.go]: you should not increment it here, it should be in the `ProcessDeposit`function

File: beacon-chain/chaintest/tests/state-tests/block-processing.yaml [hunk 6]
```diff
@@ -17,18 +17,18 @@ test_cases:
       num_slots: 64 # Testing advancing state to exactly slot == SlotsPerEpoch
       deposits:
         - slot: 9223372036854775809
-          amount: 32
-          merkle_index: 0
+          amount: 0
```
Comment (nisdas) [inline beacon-chain/chaintest/tests/state-tests/block-processing.yaml]: wait , what is going on here ? why is this 1.
the same below. We always use 32 eth

File: beacon-chain/db/pending_deposits.go (hunk 4 above)
Comment (0xKiwi) [reply beacon-chain/db/pending_deposits.go]: Good point!

**Comment by 0xKiwi** [comment]
Progress so far, I believe it's mostly done except with some bugs on the RPC side. Too late now so I'll resume tomorrow.

Also, @nisdas and I don't think the RPC should only return
```
deposit.index < current_state.deposit_index
```

Shouldn't it at least include the current deposit index, so the validators can send it?
And would the validators even need the previous deposits, doesn't it only need current and upcoming?

File: beacon-chain/rpc/beacon_server.go [hunk 7]
```diff
@@ -244,6 +243,17 @@ func (bs *BeaconServer) PendingDeposits(ctx context.Context, _ *ptypes.Empty) (*
...
+
+	// Deposits need to be received in order of merkle index root, so this has to make sure
+	// deposits are sorted from lowest to highest.
+	var pendingDeps []*pbp2p.Deposit
+	for _, dep := range allPendingDeps {
+		if dep.MerkleTreeIndex >= beaconState.DepositIndex {
+			pendingDeps = append(pendingDeps, dep)
```
Comment (nisdas) [inline beacon-chain/rpc/beacon_server.go]: this isnt ordered, we want the indexes to be sorted in an ascending order.

Comment (nisdas) [reply beacon-chain/rpc/beacon_server.go]: nvm , ignore it, its sorted by the PendingDeposits function

**Comment by rauljordan** [comment]
Looks great, awesome job on this non-trivial feature!

**Pull request merged**